app/editres/man/editres.man
'\" t
.\" Copyright 1993, 1994, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\"
.TH EDITRES 1 __xorgversion__
.SH NAME
editres \- a dynamic resource editor for X Toolkit applications
.SH SYNTAX
\fBeditres\fP [ \fI\-toolkitoption\fP .\|.\|. ] [\fB\-help\fP] [\fB\-version\fP]
.SH OPTIONS
.B Editres
accepts all of the standard X Toolkit command line options (see
.BR X (__miscmansuffix__)).
The order of the command line options is not important.
.PP
.B Editres
also accepts the following additional options:
.TP 12
.B \-help
Prints a usage message and exits.
.TP 12
.B \-version
Prints version info and exits.
.SH DESCRIPTION
.B Editres
is a tool that allows users and application developers to view
the full widget hierarchy of any X Toolkit application that speaks the
Editres protocol.
In addition,
.B editres
will help the user construct resource specifications,
allow the user to apply the resource to the application
and view the results dynamically.
Once the user is happy with a resource specification
.B editres
will append the resource string to the user's X Resources file.
.SH USING EDITRES
.B Editres
provides a window consisting of the following four areas:
.IP "Menu Bar" 25
A set of popup menus that allow you full access to \fBeditres\fP's features.
.IP "Panner"
The panner allows a more intuitive way to scroll the application tree display.
.IP "Message Area"
Displays information to the user about the action that \fBeditres\fP expects
of her.
.IP "Application Widget Tree" 25
This area will be used to display the selected application's widget tree.
.LP
To begin an \fBeditres\fP session select the \fBGet Widget Tree\fP menu item
from the command menu.
This will change the pointer cursor to a cross hair.
You should now select the application you wish look at by clicking on
any of its windows.
If this application understands the editres protocol then \fBeditres\fP
will display the application's widget tree in its tree window.
If the application does not understand the editres protocol \fBeditres\fP
will inform you of this fact in the message area after a few seconds delay.
.LP
Once you have a widget tree you may now select any of the other menu options.
The effect of each of these is described below.
.SH COMMANDS
.IP "Get Widget Tree" 8
Allows the user to click on any application that speaks the editres
protocol and receive its widget tree.
.IP "Refresh Current Widget Tree"
\fBEditres\fP only knows about the widgets that exist at the present time.
Many applications create and destroy widgets on the fly.
Selecting this menu item will cause \fBeditres\fP
to ask the application to resend its widget tree,
thus updating its information to the new state of the application.
.IP
For example, \fBxman\fP only creates the widgets for its \fItopbox\fP when it
starts up.
None of the widgets for the manual page window are created
until the user actually clicks on the \fIManual Page\fP button.
If you retrieved \fBxman\fP's widget tree before the the manual page is active,
you may wish to refresh the widget tree after the manual page has been displayed.
This will allow you to also edit the manual page's resources.
.IP "Dump Widget Tree to a File"
For documenting applications it is often useful to be able to
dump the entire application widget tree to an ASCII file.
This file can then be included in the manual page.
When this menu item is selected a popup dialog is activated.
Type the name of the file in this dialog, and either select \fIokay\fP,
or type a carriage-return.
\fBEditres\fP will now dump the widget tree to this file.
To cancel the file dialog, select the \fIcancel\fP button.
.IP "Show Resource Box"
This command will popup a resource box for the current application.
This resource box (described in detail below) will allow the user to see
exactly which resources can be set for the widget that is currently
selected in the widget tree display.
Only one widget may be currently selected;
if greater or fewer are selected \fBeditres\fP will refuse to
pop up the resource box and put an error message in the \fBMessage Area\fP.
.IP "Set Resource"
This command will popup a simple dialog box for setting an arbitrary
resource on all selected widgets.
You must type in the resource name, as well as the value.
You can use the Tab key to switch between the resource name field
and the resource value field.
.IP "Quit"
Exits
.BR editres .
.SH TREE COMMANDS
The \fBTree\fP menu contains several commands that allow operations to
be performed on the widget tree.
.IP "Select Widget in Client"
This menu item allows you to select any widget in the application; \fBeditres\fP
will then highlight the corresponding element the widget tree display.
Once this menu item is selected the pointer cursor will again turn to a
crosshair, and you must click any pointer button in the widget you
wish to have displayed.
Since some widgets are fully obscured by their children,
it is not possible to get to every widget this way,
but this mechanism does give very useful feedback between the elements
in the widget tree and those in the actual application.
.IP "Select All"
.IP "Unselect All"
.IP "Invert All"
These functions allow the user to select, unselect, or invert all
widgets in the widget tree.
.IP "Select Children"
.IP "Select Parents"
These functions select the immediate parent or children of each of the
currently selected widgets.
.IP "Select Descendants"
.IP "Select Ancestors"
These functions select all parents or children of each of the
currently selected widgets.  This is a recursive search.
.IP "Show Widget Names"
.IP "Show Class Names"
.IP "Show Widget IDs"
.IP "Show Widget Windows"
When the tree widget is initially displayed the labels of each widget
in the tree correspond to the widget names.  These functions will
cause the label of \fBall\fP widgets in the tree to be changed to show the
class name, IDs, or window associated with each widget in the application.
The widget IDs, and windows are shown as hex numbers.
.LP
In addition there are keyboard accelerators for each of the Tree operations.
If the input focus is over an individual widget in the tree,
then that operation will only effect that widget.
If the input focus is in the Tree background
it will have exactly the same effect as the corresponding menu item.
.LP
The translation entries shown may be applied to any widget in the application.
If that widget is a child of the Tree widget,
then it will only affect that widget,
otherwise it will have the same effect as the commands in the tree menu.
.IP "Flash Active Widgets"
This command is the inverse of the \fBSelect Widget in Client\fP command,
it will show the user each widget that is currently selected in
the widget tree, by flashing the corresponding widget in the
application \fInumFlashes\fP (three by default) times in the
\fIflashColor\fP.
.sp
.TS
lb lb lb
l l l.
Key	Option	Translation Entry
-
space	Unselect	Select(nothing)
w	Select	Select(widget)
s	Select	Select(all)
i	Invert	Select(invert)
c	Select Children	Select(children)
d	Select Descendants	Select(descendants)
p	Select Parent	Select(parent)
a	Select Ancestors	Select(ancestors)
N	Show Widget Names	Relabel(name)
C	Show Class Names	Relabel(class)
I	Show Widget IDs	Relabel(id)
W	Show Widget Windows	Relabel(window)
T	Toggle Widget/Class Name	Relabel(toggle)
.TE
.sp
Clicking button 1 on a widget adds it to the set of selected widgets.
Clicking button 2 on a widget deselects all other widgets and then
selects just that widget.
Clicking button 3 on a widget toggles its label between the widget's
instance name the widget's class name.
.sp
.SH USING THE RESOURCE BOX
The resource box contains five different areas.  Each of the areas,
as they appear on the screen, from top to bottom will be discussed.
.IP "The Resource Line"
This area at the top of the resource box shows the current resource
name exactly as it would appear if you were to save it to a file or
apply it.
.IP "The Widget Names and Classes"
This area allows you to select exactly which widgets this resource will
apply to.
The area contains four lines,
the first contains the name of the selected widget and all its ancestors,
and the more restrictive dot (\fB.\fP) separator.
The second line contains less specific the Class names
of each widget, and well as the less restrictive star (\fB*\fP) separator.
The third line contains a set of special buttons called \fBAny Widget\fP
which will generalize this level to match any widget.
The last line contains a set of special buttons called \fBAny
Widget Chain\fP which will turn the single level into something that
matches zero or more levels.
.sp
The initial state of this area is the most restrictive, using the
resource names and the dot separator.
By selecting the other buttons in this area you can ease the restrictions
to allow more and more widgets to match the specification.
The extreme case is to select all the \fBAny Widget Chain\fP buttons,
which will match every widget in the application.
As you select different buttons the tree display will update
to show you exactly which widgets will be effected by the current
resource specification.
.IP "Normal and Constraint Resources"
The next area allows you to select the name of the normal or
constraint resources you wish to set.
Some widgets may not have constraint resources, so that area will not appear.
.IP "Resource Value"
This next area allows you to enter the resource value.
This value should be entered exactly as you would type a line
into your resource file.
Thus it should contain no unescaped new-lines.
There are a few special character sequences for this file:
.sp
\\n - This will be replaced with a newline.
.sp
\\### - Where # is any octal digit.  This will be replaced with a
single byte that contains this sequence interpreted as an octal number.
For example, a value containing a NULL byte can be stored by
specifying \\000.
.sp
\\<new-line> - This will compress to nothing.
.sp
\\\\ - This will compress to a single backslash.
.IP "Command Area"
This area contains several command buttons, described in this section.
.IP "Set Save File"
This button allows the user to modify file that the resources will be saved to.
This button will bring up a dialog box that will ask you for a filename;
once the filename has been entered,
either hit carriage-return or click on the \fIokay\fP button.
To pop down the dialog box without changing the save file,
click the \fIcancel\fP button.
.IP "Save"
This button will append the \fBresource line\fP described above to the
end of the current save file.
If no save file has been set the \fBSet Save File\fP dialog box will be
popped up to prompt the user for a filename.
.IP "Apply"
This button attempts to perform a XtSetValues call on all widgets
that match the \fBresource line\fP described above.
The value specified is applied directly to all matching widgets.
This behavior is an attempt to give a dynamic feel to the resource editor.
Since this feature allows users to put an application in states it may not
be willing to handle,
a hook has been provided to allow specific applications to
block these SetValues requests (see \fBBlocking Editres Requests\fP below).
.sp
Unfortunately due to design constraints imposed on the widgets by the X
Toolkit and the Resource Manager, trying to coerce an inherently
static system into dynamic behavior can cause strange results.
There is no guarantee that the results of an apply will be the same as what
will happen when you save the value and restart the application.
This functionality is provided to try to give you a rough feel for what
your changes will accomplish, and the results obtained should be considered
suspect at best.
Having said that, this is one of the neatest features of editres,
and I strongly suggest that you play with it, and see what it can do.
.IP "Save and Apply"
This button combines the Save and Apply actions described above into
one button.
.IP "Popdown Resource Box"
This button will remove the resource box from the display.
.SH BLOCKING EDITRES REQUESTS
The editres protocol has been built into the Athena Widget set.
This allows all applications that are linked against Xaw to be able to speak
to the resource editor.
While this provides great flexibility, and is a useful tool,
it can quite easily be abused.
It is therefore possible for any Xaw application to specify a value for the
\fBeditresBlock\fP resource described below,
to keep \fBeditres\fP from divulging information about its internals,
or to disable the \fBSetValues\fP part of the protocol.
.TP 8
.B editresBlock (\fPClass\fB EditresBlock)
Specifies which type of blocking this application wishes to impose on the
editres protocol.
.LP
The accepted values are:
.IP all 15
Block all requests.
.IP setValues
Block all SetValues requests.
As this is the only editres request that actually modifies the application,
this is in effect stating that the application is read-only.
.IP none
Allow all editres requests.
.LP
Remember that these resources are set on any Xaw application, \fBnot editres\fP.
They allow individual applications to keep all or some
of the requests editres makes from ever succeeding.
Of course, \fBeditres\fP is also an Xaw application,
so it may also be viewed and modified by \fBeditres\fP
(rather recursive, I know), these commands can be blocked
by setting the \fBeditresBlock\fP resource on \fBeditres\fP itself.
.SH RESOURCES
For \fIeditres\fP the available application resources are:
.TP 8
.B numFlashes (\fPClass\fB NumFlashes)
Specifies the number of times the widgets in the application
will be flashed when the \fBShow Active Widgets\fP command in invoked.
.TP 8
.B flashTime (\fPClass\fB FlashTime)
Amount of time between the flashes described above.
.TP 8
.B flashColor (\fPClass\fB flashColor)
Specifies the color used to flash application widgets.  A bright color
should be used that will immediately draw your attention to the area being
flashed, such as red or yellow.
.TP 8
.B saveResourcesFile (\fPClass\fB SaveResourcesFile)
This is the file the resource line will be append to when the \fBSave\fP
button activated in the resource box.
.SH WIDGETS
In order to specify resources, it is useful to know the hierarchy of
the widgets which compose \fBeditres\fP.  In the notation below,
indentation indicates hierarchical structure.  The widget class name
is given first, followed by the widget instance name.
.sp
.\" Default tab stops will work for this display,
.\" which allows us to avoid spurious a warning duriing XML upconversion.
.nf
.TA .5i 1.0i 1.5i 2.0i
.ta .5i 1.0i 1.5i 2.0i
Editres  editres
	Paned  paned
		Box  box
			MenuButton  commands
				SimpleMenu  menu
					SmeBSB  sendTree
					SmeBSB  refreshTree
					SmeBSB  dumpTreeToFile
					SmeLine  line
					SmeBSB  getResourceList
					SmeLine  line
					SmeBSB  quit
			MenuButton  treeCommands
				SimpleMenu  menu
					SmeBSB  showClientWidget
					SmeBSB  selectAll
					SmeBSB  unselectAll
					SmeBSB  invertAll
					SmeLine  line
					SmeBSB  selectChildren
					SmeBSB  selectParent
					SmeBSB  selectDescendants
					SmeBSB  selectAncestors
					SmeLine  line
					SmeBSB  showWidgetNames
					SmeBSB  showClassNames
					SmeBSB  showWidgetIDs
					SmeBSB  showWidgetWindows
					SmeLine  line
					SmeBSB  flashActiveWidgets
		Paned  hPane
			Panner  panner
			Label  userMessage
			Grip  grip
		Porthole  porthole
			Tree  tree
				Toggle  <name of widget in application>
					.
					.
					.
					TransientShell  resourceBox
						Paned  pane
							Label  resourceLabel
							Form  namesAndClasses
								Toggle  dot
								Toggle  star
								Toggle  any
								Toggle  name
								Toggle  class
								     .
								     .
								     .
							Label  namesLabel
							List  namesList
							Label  constraintLabel
							List  constraintList
							Form  valueForm
								Label  valueLabel
								Text  valueText
							Box  commandBox
								Command  setFile
								Command  save
								Command  apply
								Command  saveAndApply
								Command  cancel
							Grip  grip
		Grip  grip
.fi
.sp
.SH ENVIRONMENT
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the \fBRESOURCE_MANAGER\fP property.
.SH FILES
.TP
.B __apploaddir__/Editres
specifies required resources
.SH SEE ALSO
.BR X (__miscmansuffix__),
.BR xrdb (__appmansuffix__),
.I Athena Widget Set
.SH RESTRICTIONS
This is a prototype, there are lots of nifty features I would love to add,
but I hope this will give you some ideas about what a resource editor can do.
.SH AUTHOR
Chris D. Peterson, formerly MIT X Consortium


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

app/editres/editres.c
/*
 *
Copyright 1989, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xaw/Cardinals.h>

#define THIS_IS_MAIN		/* Don't get extern definitions of global
				   variables. */

#include "editresP.h"

/*
 * Global variables.
 */

/* array of toolkit dependent labels taken from the resource file */
String res_labels[NUM_RES_LABELS];

/* decremented if the target client does not speak the current version */
int global_effective_protocol_version = CURRENT_PROTOCOL_VERSION;

/* toolkit type of client whose "resources" we are currently editing */
const char *global_effective_toolkit = "xt";

int global_error_code;
unsigned long global_serial_num;
int (*global_old_error_handler)(Display *, XErrorEvent *);

Boolean global_resource_box_up = FALSE;
TreeInfo *global_tree_info = NULL;
CurrentClient global_client;
ScreenData global_screen_data;
Widget global_tree_parent;
Widget global_paned = NULL;		/* named after toolkit */
Widget global_toplevel;
AppResources global_resources;


static void Syntax (XtAppContext, const char *, int) _X_NORETURN;

static String fallback_resources[] = {
    NULL,
};

#define Offset(field) (XtOffsetOf(AppResources, field))

static XtResource editres_resources[] = {
  {(String)"debug", (String)"Debug", XtRBoolean, sizeof(Boolean),
     Offset(debug), XtRImmediate, (XtPointer) FALSE},
  {(String)"numFlashes", (String)"NumFlashes", XtRInt, sizeof(int),
     Offset(num_flashes), XtRImmediate, (XtPointer) NUM_FLASHES},
  {(String)"flashTime", (String)"FlashTime", XtRInt, sizeof(int),
     Offset(flash_time), XtRImmediate, (XtPointer) FLASH_TIME},
  {(String)"flashColor", XtCForeground, XtRPixel, sizeof(Pixel),
     Offset(flash_color), XtRImmediate, (XtPointer) XtDefaultForeground},
  {(String)"saveResourceFile", (String)"SaveResourcesFile", XtRString, sizeof(String),
     Offset(save_resources_file), XtRString, (XtPointer) ""},
};

Atom wm_delete_window;

int
main(int argc, char **argv)
{
    XtAppContext app_con;

    /* Handle args that don't require opening a display */
    for (int n = 1; n < argc; n++) {
	const char *argn = argv[n];
	/* accept single or double dash for -help & -version */
	if (argn[0] == '-' && argn[1] == '-') {
	    argn++;
	}
	if (strcmp(argn, "-help") == 0) {
	    Syntax(NULL, argv[0], 0);
	}
	if (strcmp(argn, "-version") == 0) {
	    puts(PACKAGE_STRING);
	    exit(0);
	}
    }

    global_toplevel = XtAppInitialize(&app_con, "Editres", NULL, ZERO,
			       &argc, argv, fallback_resources,
			       NULL, ZERO);

    if (argc != 1) {
	fputs("Unknown argument(s):", stderr);
	for (int n = 1; n < argc; n++) {
	    fprintf(stderr, " %s", argv[n]);
	}
	fputs("\n\n", stderr);
	Syntax(app_con, argv[0], 1);
    }

    SetApplicationActions(app_con);
    XtGetApplicationResources(global_toplevel, (XtPointer) &global_resources,
			      editres_resources, XtNumber(editres_resources),
			      NULL, (Cardinal) 0);
    global_resources.allocated_save_resources_file = FALSE;

    XtOverrideTranslations
      (global_toplevel,
       XtParseTranslationTable ("<Message>WM_PROTOCOLS: quit()"));

    /* build tree for Xt intrinsics */
    BuildWidgetTree(global_toplevel);

    SetMessage(global_screen_data.info_label,
	       res_labels[13]);

    global_screen_data.set_values_popup = NULL;

    InternAtoms(XtDisplay(global_toplevel));

    XtRealizeWidget(global_toplevel);

    wm_delete_window =
      XInternAtom(XtDisplay(global_toplevel), "WM_DELETE_WINDOW",
				   False);
    (void) XSetWMProtocols (XtDisplay(global_toplevel),
			    XtWindow(global_toplevel),
                            &wm_delete_window, 1);
    XtAppMainLoop(app_con);
    exit(0);
}

/*	Function Name: Syntax
 *	Description: Prints a the calling syntax for this function to stdout.
 *	Arguments: app_con - the application context.
 *                 call - the name of the application.
 *	Returns: none - exits though.
 */

static void
Syntax(XtAppContext app_con, const char *call, int exit_val)
{
    if (app_con != NULL)
        XtDestroyApplicationContext(app_con);
    fprintf(stderr, "Usage: %s [ toolkitoptions ] [-help] [-version]\n", call);
    exit(exit_val);
}
